lib/service/rpmsg/rpc/rpmsg_rpc_server.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <metal/mutex.h>
+#include <metal/spinlock.h>
+#include <metal/utilities.h>
+#include <openamp/open_amp.h>
+#include <openamp/rpmsg_rpc_client_server.h>
+#include <string.h>
+#include <fcntl.h>
+
+#define LPRINTF(format, ...) metal_log(METAL_LOG_ERROR, format, ##__VA_ARGS__)
+
+static struct rpmsg_rpc_services table[TABLE_SIZE];
+static int table_len;
+
+static void *copy_from_shbuf(void *dst, void *shbuf, int len)
+{
+	void *retdst = dst;
+
+	while (len && ((uintptr_t)shbuf % sizeof(int))) {
+		*(uint8_t *)dst = *(uint8_t *)shbuf;
+		dst++;
+		shbuf++;
+		len--;
+	}
+	while (len >= (int)sizeof(int)) {
+		*(unsigned int *)dst = *(unsigned int *)shbuf;
+		dst += sizeof(int);
+		shbuf += sizeof(int);
+		len -= sizeof(int);
+	}
+	while (len > 0) {
+		*(uint8_t *)dst = *(uint8_t *)shbuf;
+		dst++;
+		shbuf++;
+		len--;
+	}
+	return retdst;
+}
+
+int rpmsg_rpc_server_init(struct rpmsg_rpc_services *services,
+			  int len)
+{
+	for (int i = 0; i < len; i++)
+		table[i] = services[i];
+	table_len = len;
+
+	return RPMSG_SUCCESS;
+}
+
+static struct rpmsg_rpc_services *find_service(unsigned int id)
+{
+	for (int i = 0; i < table_len; i++) {
+		if ((&table[i])->id == id) {
+			return &table[i];
+		}
+	}
+	return NULL;
+}
+
+int rpmsg_endpoint_server_cb(struct rpmsg_endpoint *ept, void *data,
+			     size_t len,
+			     uint32_t src, void *priv)
+{
+	unsigned char buf[MAX_BUF_LEN];
+	unsigned int id;
+	struct rpmsg_rpc_services *service;
+	(void)priv;
+	(void)src;
+
+	if (len > MAX_BUF_LEN)
+		return -EINVAL;
+	copy_from_shbuf(buf, data, len);
```

Comment:
no clear for me what you want to do here witjh thre function copy_from_shbuf
why not just define a rpc_msg struct 
```
struct rpmsg_rpc {
	unsigned int id;
	size_t param_size; /* optional can be deduced from the rpmsg size */
        uint8_t param[0] 
};

```
and then you have just to cast the message 

```
	struct rpmsg_rpc *msg = (struct rpmsg_rpc *) data;

	service = find_service(msg->id);
	if (service) {
		service->respcb.cb_function(msg->param, msg->param_size, ept);

```